R/combineResults.R

```diff
@@ -0,0 +1,232 @@
+#'     colData=DataFrame(label=g)
+#' )
+#' 
+#' rownames(test) <- sprintf("GENE_%s", seq_len(nrow(test)))
+#' test <- scater::logNormCounts(test)
+#' 
+#' ###############################
+#' ## Performing classification ##
+#' ###############################
+#' 
+#' pred1 <- SingleR(test, ref1, labels=ref1$label)
+#' pred2 <- SingleR(test, ref2, labels=ref2$label)
+#'
+#' pred3 <- SingleR(test, ref1, labels=ref1$label, 
+#'     method="cluster", clusters=test$label) 
+#' pred4 <- SingleR(test, ref2, labels=ref2$label, 
+#'     method="cluster", clusters=test$label) 
+#'
+#' ###############################
+#' ##     Combining results     ##
+#' ###############################
+#'
+#' pred.single <- combineResults(list(pred1, pred2))
+#' pred.clust <- combineResults(list(pred3, pred4))
+#'
+#' @seealso
+#' \code{\link{matchReferences}}, to harmonize labels between reference datasets.
+#' \code{\link{SingleR}}, for generating predictions.
+#'
+#' @importFrom S4Vectors DataFrame metadata metadata<-
+#'
+#' @export
+combineResults <- function(results) {
+    num.features <- sapply(results, nrow)
+    if (abs(max(num.features) - min(num.features)) != 0) {
+        stop("Results DataFrames contain different numbers of cells or clusters.")
+    }
+
+    pred.scores <- list()
+    pred.firstlabels <- list()
+    pred.prunedlabels <- list()
+    pred.tuningscores <- list()
+    feature.names <- list()
+    de.genes <- list()
+    common.genes <- list()
+
+    # Keep track of which scores belong to which results object.
+    pred.score.indices <- list()
+    ind <- 1
+
+    for (i in seq_along(results)) {
+        res <- results[[i]]
+        pred.scores[[i]] <- res$scores
+
+        if (!is.null(res$first.labels)) {
+            pred.firstlabels[[i]] <- res$first.labels
+        } else {
+            pred.firstlabels[[i]] <- NA_character_
+        }
+
+        if (!is.null(res$pruned.labels)) {
+            pred.prunedlabels[[i]] <- res$pruned.labels
+        } else {
+            pred.prunedlabels[[i]] <- NA_character_
+        }
+
+        if (!is.null(res$tuning.scores)) {
+            pred.tuningscores[[i]] <- res$tuning.scores
+        } else {
+            pred.tuningscores[[i]] <- NA_character_
+        }
+
+        feature.names[[i]] <- rownames(res)
+
+        if (!is.null(metadata(res)$de.genes)) {
+            de.genes[[i]] <- metadata(res)$de.genes
+        } else {
+            de.genes[[i]] <- NA_character_
+        }
+
+        common.genes[[i]] <- metadata(res)$common.genes
+        pred.score.indices[[i]] <- seq(ind, (ind + ncol(res$scores) - 1))
+        ind <- ind + ncol(res$scores)
+    }
+
+    all.scores <- do.call(cbind, pred.scores)
+    best.labels <- colnames(all.scores)[max.col(all.scores)]
+
+    # Determine which results DataFrame each score comes from.
+    pred.indices <- sapply(max.col(all.scores), .get_pred_indices, pred.score.indices = pred.score.indices)
```

Comment:
Kind of a roundabout way of doing this; you're calling out to another function, which makes another loop, and you also have to keep track of `pred.score.indices`. I would do it more like the pseudo-code below (note, untested):

```r
    ncells <- nrow(results[[1]]) # assume that you've put in checks for the same num & id of rows

    last.best <- rep(-Inf, ncells) 
    chosen.label <- chosen.first <- chosen.pruned <- rep(NA_character_, ncells)
    collected.scores <- list()

    for (i in seq_along(results)) {
        res <- results[[i]]
        curbest <- scores[cbind(seq_len(ncells), max.col(scores))]
        collected.scores[[i]] <- curbest
        better <- curbest > last.best
         
        chosen.label[better] <- res$labels[better]
 
        if (has.first) { # either everyone has 'first', or no-one does.
            chosen.first[better] <- res$first.labels[better]
        }

        if (has.pruned) { # same for pruned.
             chosen.pruned[better] <- res$pruned.labels[better]
        }
    }

    all.scores <- do.call(cbind, collected.scores)
```

Reply:
So this is snagging the label from the result with the highest score, rather than the actual label of the highest score (which is how it currently works). Is that the functionality we want?

Reply:
That's correct. Even with a single reference, we don't actually get the label with the highest score anyway after fine-tuning, so it's no big deal that we're not doing that with multiple references.